File: src/languages/marathi/marathi.ts
/**
 * Native Marathi (mr) text phonemizer — canonical IPA. Marathi is written in Devanagari
 * and shares almost all of Hindi's abugida machinery, so it REUSES the generic Hindi engine (makeNativeHindi)
 * with a Marathi data file (marathi.jsonc). The Marathi-specific facts live entirely in the manifest:
 *   - ळ → retroflex lateral [ɭ]; ष → retroflex [ʂ] (Hindi merges it to ʃ);
 *   - च/छ/ज/झ → DENTAL affricates [t͡s t͡sʰ d͡z d͡zʱ] before a back/central vowel (postRules), palatal before front;
 *   - no Hindi और-offglide / əɦə→ɛɦɛ finalRules (Marathi keeps शहर→ɕəɦəɾ);
 *   - Marathi number spellings.
 * Inherent-schwa deletion + weight stress are the same shared algorithms as Hindi.
 */
import { makeNativeHindi, type HindiDef, type ForeignPhonemizer } from "../hindi/hindi.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { loadSharedPhonology } from "../../core/phonology.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { makeMarathiNormalizer, type MarathiWords } from "./normalize.ts";

/**
 * Marathi's manifest adds the word tables Hindi's def does not carry. ⚠ `MarathiWords` is normalize.ts's
 * half; the tier below reads `percent`, `currency`, `ampersand`, `multiply` and `units` from the SAME
 * object, which is the whole point — see the £ note in marathi.jsonc.
 */
export interface MarathiDef extends HindiDef, MarathiWords {
    ampersand: string;
    multiply: string;
    units: Record<string, string>;
}

// ⚠ ONE LOAD, MODULE-LEVEL. The tier below is built at import time and must read the same object the engine
// does, or the two go back to disagreeing — which is exactly what happened to £ (see marathi.jsonc).
export const DEF = loadManifest<MarathiDef>(import.meta.url, "marathi.jsonc");

// The bare-sign map the Hindi engine tokenizes on is DERIVED from `percent`, not authored a second time.
const WITH_SYMBOLS: MarathiDef = { ...DEF, symbols: { ...(DEF.symbols ?? {}), "%": DEF.percent.plural } };


let MR: ReturnType<typeof makeNativeHindi> | undefined;
function engine(foreign?: ForeignPhonemizer) {
    return makeNativeHindi(
        WITH_SYMBOLS,
        loadSharedPhonology(),
        foreign,
    );
}

/** Normalization. Marathi shares Hindi's ENGINE but not Hindi's orthographic conventions, so it supplies
 *  its OWN normalizer and its OWN symbol words through `makeNativeHindi`'s overrides rather than
 *  inheriting Hindi's. */
const MR_SYMBOLS = makeSymbolNormalizer({
    // `multiply` — this language had NO word for the sign at all. ⚠ STANDARD MATHEMATICAL REGISTER, not a
    // corpus attestation: the sweep's plausible hits were homographs of PREPOSITIONS (es `por` ×23, it `per` ×25,
    // ru `на` ×31 are all the preposition), the same trap that defeated the exponent sourcing. One word, so `by`
    // defaults to it — this language does not split dimension from product.
    multiply: { times: DEF.multiply },
    // `&` was DROPPED outright: the corpus's `B&B` and `Arts & Sciences` lost the sign.
    // `आणि` ×1073 in this corpus. The tier spaces it on both sides, because `B&B` is two
    ampersand: DEF.ampersand,
    percent: [DEF.percent.plural], // Hindi's प्रतिशत is not Marathi
    // From the manifest, and shared with normalize.ts — the two paths claim the sign in different positions
    // and used to answer with different words for £.
    currency: Object.fromEntries(Object.entries(DEF.currency).map(([k, v]) => [k, [v]])),
    // ⚠ ⟨ha⟩ ⟨l⟩ ⟨L⟩ WERE MIS-READING, NOT LEAKING — `10 ha` read *d̪ˈəɦaː hˈɑː* and `10 l` *d̪ˈəɦaː ˈɛɫ*,
    // the English letter name out of a Devanagari engine, with no ASCII surviving and nothing vanishing.
    // See `tools/normalization/misread.ts`. Each word is definitional on mr.wikipedia:
    //   हेक्टर  52/13  "हेक्टर हे क्षेत्रफळ मोजण्याचे एकक आहे … १०० मीटर X १०० मीटर = १ हेक्टर = १०००० वर्ग मीटर"
    //   लिटर    26/12  "…डब्यात मावणाऱ्या द्रवाबरोबरचे आकारमान हे लिटर होय. तसेच, १००० मिली लिटर = १ लिटर"
    // ⚠ हेक्टर IS THE HECTARE IN MARATHI AND HECTOR IN HINDI — the same string, opposite verdicts, and the
    // reason both were read rather than counted. hi's हेक्टर is 56 tokens of the Trojan prince and its
    // hectare is हेक्टेयर; mr's हेक्टर heads the HECTARE article, and its personal-name hits (हेक्टर मोरेनो
    // the footballer) are the minority sense. Neither language's answer could be borrowed from the other.
    // ⚠ ⟨g⟩ IS REFUSED THOUGH ग्रॅम IS THE BEST-ATTESTED WORD HERE (88 tokens / 20 arts, definitional:
    // "किलोग्रॅम हे वजनाचे एकक आहे … याचे एस. आय. संक्षिप्त नाम kg आहे", and "१० ग्रॅम शुद्ध सोन्याचा" is a
    // digit-adjacent gram). The artifact's only `<digit> g` is `802.11 g` — a Wi-Fi standard — and it is
    // written WITH A SPACE, which the tier's `NOT_VERSION` guard cannot see: that guard requires the letter
    // GLUED to the number (`\d+[.,]\d+[a-zA-Z]`), deliberately, because `12.5 g` is a real measurement of
    // exactly the spaced shape. Verified against a language that does declare ⟨g⟩: `802.11 g` reads
    // *… ɛlf ɡʁam* in de and *… wˈʌn ɡɹˈæmz* in en. Undecidable at this layer, so mr declines the key.
    // ⚠ ⟨m⟩ stays refused for the reason `normalize.ts` already records — `100m`/`200m` are swim events.
    // From the manifest, shared with normalize.ts's own (larger) table — marathi.jsonc records why the
    // two tables exist and must not be merged.
    units: Object.fromEntries(Object.entries(DEF.units).map(([k, v]) => [k, [v]])),
});

export function createMarathi(foreign?: ForeignPhonemizer): {
    text(input: string): string;
} {
    return makeNativeHindi(WITH_SYMBOLS, loadSharedPhonology(), foreign, undefined, undefined, {
        normalize: makeMarathiNormalizer(DEF),
        symbols: MR_SYMBOLS,
    });
}

/** Bare word→IPA (tests / referee eval). */
export function phonemizeWord(w: string): string {
    return (MR ??= engine()).word(w);
}

File: csharp/Vernacula.Phonemizer/Languages/Marathi/Marathi.cs
/**
 * Native Marathi (mr) text phonemizer — canonical IPA. Reuses the generic Hindi engine (MakeNativeHindi) with
 * a Marathi manifest; the Marathi-specific facts live entirely in marathi.jsonc.
 * Ported from src/languages/marathi/marathi.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;
using Vernacula.Phonemizer.Languages.Hindi;

namespace Vernacula.Phonemizer.Languages.Marathi;

/** Marathi's manifest adds the two word tables Hindi's def does not carry. */
public sealed class MarathiDef : HindiDef
{
    public PercentDef Percent { get; set; } = new();
    public Dictionary<string, string> Currency { get; set; } = new();
    public string Ampersand { get; set; } = "";
    public string Multiply { get; set; } = "";
    /** The LATIN unit abbreviations the shared tier is keyed on. */
    public Dictionary<string, string> Units { get; set; } = new();
    /** ⚠ A SECOND unit table on purpose — normalize.ts's, with the Devanagari forms. See marathi.jsonc. */
    public Dictionary<string, string> UnitWords { get; set; } = new();
    public string[] MagnitudeWords { get; set; } = [];
    public OrdinalsDef Ordinals { get; set; } = new();
    public string[] VisargaAdverbs { get; set; } = [];
    public ClockDef Clock { get; set; } = new();
    public EraDef EraMarkers { get; set; } = new();
    public Dictionary<string, string> Abbreviations { get; set; } = new();
    public DegreeDef Degree { get; set; } = new();
    public FractionsDef Fractions { get; set; } = new();
    public SymbolWordsDef SymbolWords { get; set; } = new();
    public string RangeWord { get; set; } = "";
    public string BareHundred { get; set; } = "";

    public sealed class OrdinalsDef
    {
        public Dictionary<string, int> SuffixForm { get; set; } = new();
        /** ⚠ Indexed [masc, fem, plural/neuter, oblique] — the order IS the contract with SuffixForm. */
        public Dictionary<string, string[]> Irregular { get; set; } = new();
        public string StemHundred { get; set; } = "";
        public string[] StemNine { get; set; } = [];
        public string[] StemTens { get; set; } = [];
    }
    public sealed class ClockDef { public string Past { get; set; } = ""; public string Minutes { get; set; } = ""; public string Oclock { get; set; } = ""; }
    public sealed class EraDef { public string Bc { get; set; } = ""; public string Ad { get; set; } = ""; }
    public sealed class DegreeDef
    {
        public string Word { get; set; } = ""; public string Celsius { get; set; } = ""; public string Fahrenheit { get; set; } = "";
        public string North { get; set; } = ""; public string South { get; set; } = ""; public string East { get; set; } = ""; public string West { get; set; } = "";
    }
    public sealed class FractionsDef
    {
        public string Half { get; set; } = ""; public string Quarter { get; set; } = "";
        public string ThreeQuarters { get; set; } = ""; public string DividedBy { get; set; } = "";
    }
    public sealed class SymbolWordsDef
    {
        public string Plus { get; set; } = ""; public string Approximately { get; set; } = ""; public string PlusMinus { get; set; } = "";
        public string LessThan { get; set; } = ""; public string GreaterThan { get; set; } = ""; public string Divide { get; set; } = ""; public string Equals { get; set; } = "";
        /** Declared, but no ASCII-hyphen rule reads it — see marathi.jsonc. */
        public string Minus { get; set; } = "";
    }

    public sealed class PercentDef
    {
        public string Plural { get; set; } = "";
        public string Singular { get; set; } = "";
    }
}

public static class MarathiPhonemizer
{
    // ⚠ ONE LOAD. The tier below is built at type-init and must read the same object the engine does, or the
    // two go back to disagreeing — which is exactly what happened to £ (see marathi.jsonc).
    public static readonly MarathiDef DEF = LoadManifest.Load<MarathiDef>("languages/marathi", "marathi.jsonc");

    // The bare-sign map the Hindi engine tokenizes on is DERIVED from `percent`, not authored a second time.
    private static readonly MarathiDef WITH_SYMBOLS = WithPercentSign(DEF);

    private static MarathiDef WithPercentSign(MarathiDef d)
    {
        d.Symbols = new Dictionary<string, string>(d.Symbols ?? new Dictionary<string, string>(), StringComparer.Ordinal)
        {
            ["%"] = d.Percent.Plural,
        };
        return d;
    }

    private static NativeHindiEngine? MR;
    private static NativeHindiEngine Engine(ForeignPhonemizer? foreign = null) =>
        Hindi.Hindi.MakeNativeHindi(WITH_SYMBOLS, PhonologyLoader.LoadSharedPhonology(), foreign);

    /**
     * Marathi shares Hindi's ENGINE but not Hindi's orthographic conventions, so it supplies its OWN symbol
     * words (and its own normalizer) through MakeNativeHindi's overrides rather than inheriting Hindi's.
     */
    private static readonly Func<string, string> MR_SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        Multiply = new MultiplyDef { Times = "गुणिले" },
        Ampersand = "आणि",
        Percent = new[] { DEF.Percent.Plural }, // Hindi's प्रतिशत is not Marathi
        // From the manifest, and shared with Normalize.cs — the two paths claim the sign in different
        // positions and used to answer with different words for £.
        Currency = DEF.Currency.ToDictionary(kv => kv.Key, kv => (IReadOnlyList<string>)new[] { kv.Value }, StringComparer.Ordinal),
        Units = new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal)
        {
            ["km"] = new[] { "किलोमीटर" }, ["cm"] = new[] { "सेंटीमीटर" }, ["mm"] = new[] { "मिलिमीटर" },
            ["kg"] = new[] { "किलोग्रॅम" }, ["ha"] = new[] { "हेक्टर" }, ["l"] = new[] { "लिटर" },
            // ⟨l⟩ and ⟨L⟩ are both official for the litre — the one exception to the one-letter case rule in
            // Core/NormalizeSymbols, which exists for symbols whose two cases are DIFFERENT units.
            ["L"] = new[] { "लिटर" },
        },
    });

    public static NativeHindiEngine CreateMarathi(ForeignPhonemizer? foreign = null)
    {
        return Hindi.Hindi.MakeNativeHindi(WITH_SYMBOLS, PhonologyLoader.LoadSharedPhonology(), foreign, null, null,
            normalizeOverride: Normalize.MakeMarathiNormalizer(DEF),
            symbolsOverride: MR_SYMBOLS);
    }

    /** Bare word→IPA (tests / referee eval). */
    public static string PhonemizeWord(string w) => (MR ??= Engine()).Word(w);

    internal static void RegisterSelf() =>
        Registry.Register("marathi", () => new NativeHindiLanguage(CreateMarathi(latin => Registry.ReadAsEnglish(latin))));
}
